platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/NodeId.java

```diff
@@ -16,148 +16,73 @@
+ *
+ * @param id ID number unique within the network
  */
-public class NodeId implements EventCreationRule {
-    /** used to distinguish between a main node and a mirror node */
-    private boolean isMirror;
-    /** ID number unique within the network, unique set for main network and mirror network */
-    private long id;
+public record NodeId(long id) implements Comparable<NodeId> {
+
+    /** the map of all NodeId objects created, indexed by ID number */
+    private static final ConcurrentHashMap<Long, NodeId> nodeIds = new ConcurrentHashMap<>();
+
+    /** an invalid NodeId used in testing and boundary conditions. */
+    public static final NodeId INVALID_NODE_ID = NodeId.create(-1L);
+
+    /** the first NodeId object created */
+    public static final NodeId FIRST_NODE_ID = NodeId.create(0L);
 
     /**
      * Constructs a NodeId object
      *
-     * @param isMirror
-     * 		is it a mirror node or main node
-     * @param id
-     * 		the ID number
+     * @param id the ID number
      */
-    public NodeId(boolean isMirror, long id) {
-        this.isMirror = isMirror;
-        this.id = id;
+    public NodeId {
+        if (id < -1L) {
+            throw new IllegalArgumentException("id must be non-negative");
+        }
     }
 
     /**
      * Constructs a main network NodeId object
      *
-     * @param id
-     * 		the ID number
-     * @return the object created
-     */
-    public static NodeId createMain(long id) {
-        return new NodeId(false, id);
-    }
-
-    /**
-     * Constructs a mirror network NodeId object
-     *
-     * @param id
-     * 		the ID number
+     * @param id the ID number
      * @return the object created
      */
-    static NodeId createMirror(long id) {
-        return new NodeId(true, id);
+    public static NodeId create(long id) {
+        return nodeIds.computeIfAbsent(id, NodeId::new);
     }
 
     /**
-     * Checks if two IDs belong to the same network
+     * Checks if this NodeId is equal to the given ID value
      *
-     * @param nodeId
-     * 		the NodeId to compare to
-     * @return true if networks are the same, false if not
+     * @param id the ID value to compare
+     * @return true if this ID value is equal to the supplied value, false otherwise
      */
-    public boolean sameNetwork(NodeId nodeId) {
-        return this.isMirror() == nodeId.isMirror();
-    }
-
-    private boolean equals(boolean isMirror, long id) {
-        return this.isMirror() == isMirror && id == this.getId();
+    public boolean matches(long id) {
+        return this.id == id;
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
-    public boolean equals(Object obj) {
-        if (!(obj instanceof NodeId)) {
-            throw new IllegalArgumentException("obj must be a NodeId object");
-        }
-        return equals((NodeId) obj);
-    }
-
-    @Override
-    public int hashCode() {
-        return Boolean.hashCode(isMirror) + Long.hashCode(id);
```

Comment:
Your PR body says "Removing the hashcode implementation causes the app to misbehave, so it is left in.", but it's been deleted

Reply:
old PR text.  I've stricken the text from the PR.   Testing showed the record's hash function was sufficient and the problem I encountered was caused by something different.